### PR TITLE
Add a cycle method that repeats a dataset indefinitely

### PR DESCRIPTION
**Pull request by thequilo**

### CODE REVIEW

**Comment by boeddeker** [comment]
Does this dataset work with reshuffle and prefetch?

What was the motivation for the name? The cycler from matplotlib?

**Comment by thequilo** [comment]
This works with reshuffle, but it doesn't currently work with prefetch. That is because prefetch currently needs the length of the input dataset. We can make it work with prefetch, but I'm not sure if we can make it work with prefetch and reshuffle. The motivation for the name comes from [Rust's iterator](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.cycle). An alternative would be `dataset.tile(None)` or `dataset.tile(-1)`.

**Comment by boeddeker** [comment]
I checked Python itertools, it has also a cycle function. So the name is fine.

Could you move the method to be below the tile function? We should try to sort the code, that the method order is not completely random.

I don't like, that it doesn't work with reshuffle and prefetch.
At the moment I am not sure, how to get this working. 
Does it currently fail, or does it do a fallback to single shuffle?
We should at least document this.

Btw. what is the motivation for the infinity repeat?

**Comment by thequilo** [comment]
The `itertools.cycle` function is slightly different: It caches the elements of the first iteration and re-uses them in following iterations.

I also don't like that it doesn't work with shuffle and prefetch, but that is not so easy to implement because prefetch needs an indexible dataset. Prefetch currently fails because it checks the length which is not implemented for `CycleDataset`.

The motivation for this comes from `mms_msg` when somebody wants to use dynamic mixing but not our trainer.

**Comment by boeddeker** [comment]
> The itertools.cycle function is slightly different: It caches the elements of the first iteration and re-uses them in following iterations.

Yes, the caching is different, but I like it, when we use names, that are already used in python. So the name is fine. (Some other methods here also differ slightly from itertools (e g. groupby follows pandas implementation.)

> I also don't like that it doesn't work with shuffle and prefetch, but that is not so easy to implement because prefetch needs an indexible dataset. Prefetch currently fails because it checks the length which is not implemented for CycleDataset.

We should document, that cycle should be used after the reshuffle and prefetch, which has only a minor performance impact. And adding a ToDo, that we should try to find a way to get it working with reshuffle before and prefetch after (Using classical iterators instead of generators may solve this, but requires a mayor refactoring of everything).

> The motivation for this comes from mms_msg when somebody wants to use dynamic mixing but not our trainer. We specifically hit the case where we wanted to sample dialogs as part of a more complex scenario (audio event detection) where the scene sampler wants an infinite stream of dialogs.

Strange, usually each training loop implementation support somehow the concept of epochs, where you have a finite iterable (This is not a special property of our trainer).

I think the sence sampler should also accept a finite length as input. An infinite length input is not common and it is easy to support this. But this is a discussion for mms_msg.

Nonetheless, the concept of infinitly long datasets should be supported in our Datasets. We should just document, that some parts don't work. The cycle also simplifies the code of a simple training loop.

**Review by boeddeker — APPROVED** [review]
